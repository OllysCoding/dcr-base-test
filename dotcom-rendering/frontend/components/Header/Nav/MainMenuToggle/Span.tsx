import React from 'react';

export const Span: React.SFC<{
    count: number;
}> = ({ count }) => <span>Clicked {count} times</span>;
