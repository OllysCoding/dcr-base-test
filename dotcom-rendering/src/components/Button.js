// @flow
import styled from 'preact-emotion';

const pillarColours = {
    news: 'pink',
    alex: 'hotpink',
};

export default styled('button')(({ pillar = 'alex' }) => ({
    border: 'none',
    backgroundColor: pillar ? pillarColours[pillar] : null,
}));
