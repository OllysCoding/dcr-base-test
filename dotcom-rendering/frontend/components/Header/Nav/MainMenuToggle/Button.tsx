import React, { Component } from 'react';
import { css, cx } from 'react-emotion';

import { Span } from './Span';

const button = css`
    font-size: 10px;
    color: red;
`;

const focus = css`
    color: blue;
`;

interface Props {
    inFocus: boolean;
    onMouseEnter: () => void;
    onMouseLeave: () => void;
}

export class Button extends Component<Props, { count: number }> {
    constructor(props: Props) {
        super(props);

        this.state = {
            count: 0,
        };
    }

    public doSomethingToState() {
        const currentCount = this.state.count;

        this.setState({
            count: currentCount + 1,
        });
    }

    public render() {
        const { inFocus, onMouseEnter, onMouseLeave } = this.props;
        return (
            <button
                className={cx(button, {
                    [focus]: inFocus,
                })}
                onClick={this.handleClick}
                onMouseEnter={onMouseEnter}
                onMouseLeave={onMouseLeave}
            >
                <Span count={this.state.count} />
            </button>
        );
    }

    public handleClick: () => void = () => {
        const currentCount = this.state.count;
        this.setState({
            count: currentCount + 1,
        });
    };
}
