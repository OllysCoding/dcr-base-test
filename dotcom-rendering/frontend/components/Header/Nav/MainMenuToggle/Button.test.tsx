import React from 'react';
import { shallow, mount } from 'enzyme';
import { Button } from './Button';
import { Span } from './Span';

describe('Button', () => {
    let onMouseEnter: () => void;
    let onMouseLeave: () => void;

    beforeEach(() => {
        onMouseEnter = jest.fn();
        onMouseLeave = jest.fn();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    describe('snapshots', () => {
        it('It should render with correct styles if inFocus is false', () => {
            const component = shallow(
                <Button
                    inFocus={false}
                    onMouseEnter={onMouseEnter}
                    onMouseLeave={onMouseLeave}
                />,
            );

            expect(component).toMatchSnapshot();
        });

        it('It should render with correct styles if inFocus is true', () => {
            const component = shallow(
                <Button
                    inFocus={true}
                    onMouseEnter={onMouseEnter}
                    onMouseLeave={onMouseLeave}
                />,
            );

            expect(component).toMatchSnapshot();
        });

        it('It should update when clicked', () => {
            const component = shallow(
                <Button
                    inFocus={false}
                    onMouseEnter={onMouseEnter}
                    onMouseLeave={onMouseLeave}
                />,
            );

            component.simulate('click');

            expect(component).toMatchSnapshot();
        });
    });

    it('It should call onMouseEnter prop when mouseEnter event occurs', () => {
        const component = shallow(
            <Button
                inFocus={false}
                onMouseEnter={onMouseEnter}
                onMouseLeave={onMouseLeave}
            />,
        );

        component.simulate('mouseEnter');

        expect(onMouseEnter).toHaveBeenCalled();
    });

    it('It should call onMouseLeave prop when mouseLeave event occurs', () => {
        const component = shallow(
            <Button
                inFocus={false}
                onMouseEnter={onMouseEnter}
                onMouseLeave={onMouseLeave}
            />,
        );

        component.simulate('mouseLeave');

        expect(onMouseLeave).toHaveBeenCalled();
    });

    it('It should update state when clicked', () => {
        const component = shallow(
            <Button
                inFocus={false}
                onMouseEnter={onMouseEnter}
                onMouseLeave={onMouseLeave}
            />,
        );

        expect(component.state('count')).toBe(0);

        component.simulate('click');

        expect(component.state('count')).toBe(1);
    });

    it('It should render child component <Span>', () => {
        const component = mount(
            <Button
                inFocus={false}
                onMouseEnter={onMouseEnter}
                onMouseLeave={onMouseLeave}
            />,
        );

        expect(component.find(Span).length).toBe(1);
        expect(
            component
                .find(Span)
                .first()
                .prop('count'),
        ).toBe(0);
    });

    it('It should pass correct props child component <Span> after click', () => {
        const component = mount(
            <Button
                inFocus={false}
                onMouseEnter={onMouseEnter}
                onMouseLeave={onMouseLeave}
            />,
        );

        expect(
            component
                .find(Span)
                .first()
                .prop('count'),
        ).toBe(0);

        component.simulate('click');

        expect(
            component
                .find(Span)
                .first()
                .prop('count'),
        ).toBe(1);
    });

    it('It should doSomethingToState', () => {
        const component = shallow(
            <Button
                inFocus={false}
                onMouseEnter={onMouseEnter}
                onMouseLeave={onMouseLeave}
            />,
        );

        const componentInstance = component.instance() as Button;

        expect(
            component
                .find(Span)
                .first()
                .prop('count'),
        ).toBe(0);

        componentInstance.doSomethingToState();

        expect(
            component
                .find(Span)
                .first()
                .prop('count'),
        ).toBe(1);
    });
});
