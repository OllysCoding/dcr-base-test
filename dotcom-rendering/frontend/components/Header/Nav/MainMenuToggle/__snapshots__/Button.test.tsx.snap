// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshots It should render with correct styles if inFocus is false 1`] = `
<button
  className="css-168vu8a"
  onClick={[Function]}
  onMouseEnter={[MockFunction]}
  onMouseLeave={[MockFunction]}
>
  <Component
    count={0}
  />
</button>
`;

exports[`Button snapshots It should render with correct styles if inFocus is true 1`] = `
<button
  className="css-vam8e5"
  onClick={[Function]}
  onMouseEnter={[MockFunction]}
  onMouseLeave={[MockFunction]}
>
  <Component
    count={0}
  />
</button>
`;

exports[`Button snapshots It should update when clicked 1`] = `
<button
  className="css-168vu8a"
  onClick={[Function]}
  onMouseEnter={[MockFunction]}
  onMouseLeave={[MockFunction]}
>
  <Component
    count={1}
  />
</button>
`;
